syntax = "proto3";

package tunnelbroker;

import "google/protobuf/empty.proto";

service TunnelBrokerService {
  rpc NewSession(NewSessionRequest) returns (NewSessionResponse) {}
  rpc Send(SendRequest) returns (google.protobuf.Empty) {}
  rpc Get(GetRequest) returns (stream GetResponse) {}
}

// Session

message NewSessionRequest {
  string deviceID = 1;
  string notifyToken = 2;
  enum deviceType {
    MOBILE = 0;
    WEB = 1;
    KEYSERVER = 2;
  }
  message deviceInfo {
    string appVersion = 1;
    deviceType type = 2;
    string os = 3;
  }
}

message NewSessionResponse {
  string sessionID = 1;
}

// Send payload to device

message SendRequest {
  string sessionID = 1;
  string toDeviceID = 2;
  bytes payload = 3;
  repeated string blobHash = 4;
}

// Get messages from devices

message GetRequest {
  string sessionID = 1;
}

message GetResponse {
  string deviceID = 1;
  bytes payload = 2;
  repeated string blobHash = 3;
}
